Strata/DL/HighStrata/HighStrata.lean

```diff
@@ -0,0 +1,245 @@
+/-
+  Copyright Strata Contributors
+
+  SPDX-License-Identifier: Apache-2.0 OR MIT
+-/
+
+/-
+The high strata language is supposed to serve as an intermediate verification language for at least Java, Python, JavaScript.
+
+It enables doing various forms of verification:
+- Deductive verification
+- Property based testing
+- Data-flow analysis
+
+Features currently not present:
+- Type inference. The source program needs to specify enough types so that no inference is needed.
+- Type checking. We assume types have already been checked before.
+- Namespaces. All definition and reference names consist of a single Identifier
+
+Design choices:
+- Pure contracts: contracts may only contain pure code. Pure code does not modify the heap, neither by modifying existing objects are creating new ones.
+- Callables: instead of functions and methods we have a single more general concept called a 'callable'.
+- Purity: Callables can be marked as pure or impure. Pure callables have a reads clause while impure ones have a modifies clause.
+  A reads clause is currently not useful for impure callables, since reads clauses are used to determine when the output changes, but impure callables can be non-determinismic so the output can always change.
+- Opacity: callables can have a body that's transparant or opaque. Only an opaque body may declare a postcondition. A transparant callable must be pure.
+- StmtExpr: Statements and expressions are part of the same type. This reduces duplication since the same concepts are needed in both, such as conditions and variable declarations.
+- Loops: The only loop is a while, but this can be used to compile do-while and for loops to as well.
```

Comment:
This is great, as long as we have some way to recover the higher-level loop (e.g.., to recognize a "foreach" loop that can be vectorized), but I take this as more the job the Strata framework than an individual dialect.

Reply:
This PR is only intended as a first version. In particular, the current set of constructs is what's needed to support a JVerify deductive verification use-case. Even if it would be impossible to recover the higher-level loop, we can still add one later so the recovery is not needed.